pkg/execution/state/redis_state/key_generator.go

```diff
@@ -44,119 +42,105 @@ type KeyGenerator interface {
 	// for given workflow run.
 	Errors(context.Context, state.Identifier) string
 
-	// PauseLease stores the key which references a pause's lease.
-	//
-	// This is stored independently as we may store more than one copy of a pause
-	// for easy iteration.
-	PauseLease(context.Context, uuid.UUID) string
-
-	// PauseID returns the key used to store an individual pause from its ID.
-	PauseID(context.Context, uuid.UUID) string
-
-	// PauseEvent returns the key used to store data for loading pauses by events.
-	PauseEvent(context.Context, uuid.UUID, string) string
-
-	// PauseStep returns the prefix of the key used within PauseStep.  This lets us
-	// iterate through all pauses for a given identifier
-	PauseStepPrefix(context.Context, state.Identifier) string
-
-	// PauseStep returns the key used to store a pause ID by the run ID and step ID.
-	PauseStep(context.Context, state.Identifier, string) string
-
-	// PauseIndex is a key that's used to index added/expired times for pauses.
-	//
-	// Added times are necessary to load pauses after a specific point in time,
-	// which is used when caching pauses in-memory to only load the subset of pauses
-	// added after the cache was last updated.
-	PauseIndex(ctx context.Context, kind string, wsID uuid.UUID, event string) string
-
-	// RunPauses stores pause IDs for each run as a zset
-	RunPauses(ctx context.Context, runID ulid.ULID) string
-
-	// Invoke returns the key used to store the correlation key associated with invoke functions
-	Invoke(ctx context.Context, wsID uuid.UUID) string
-
 	// History returns the key used to store a log entry for run hisotry
 	History(ctx context.Context, runID ulid.ULID) string
 
 	// Stack returns the key used to store the stack for a given run
 	Stack(ctx context.Context, runID ulid.ULID) string
 }
 
-type DefaultKeyFunc struct {
-	Prefix string
+type runStateKeyGenerator struct {
+	stateDefaultKey string
 }
 
-func (d DefaultKeyFunc) Idempotency(ctx context.Context, id state.Identifier) string {
-	return fmt.Sprintf("%s:key:%s", d.Prefix, id.IdempotencyKey())
+func (s runStateKeyGenerator) Prefix(defaultPrefix string, runID ulid.ULID) string {
+	if IsSharded(runID) {
```

Comment:
Some clients and their key generators support sharding, so the keys will include a dynamic hash tag as well. This way, we will assign keys to cluster slots uniformly.